### PR TITLE
Replace crypto ops

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This takes care of replacing the `crypto_ops` struct as requested in https://github.com/OP-TEE/optee_os/pull/1908

But this doesn't solve the problem of supplying AES-GCM from an OP-TEE internal implementation and AES-CCM from LTC. How can that be done? `core/tee/tee_svc_cryp.c` expects `crypto_authenc_*()` functions so those need to be provided somewhere, but not in LTC as `crypto_authenc_*()` implementation will use OP-TEE internal AES-GCM implementation further on.

One way forward could be:
1. Specify `crypto_aes_gcm_*()` and `crypto_aes_ccm_*()` in `core/include/crypto/aes-{g,c}cm.h`
2. Move the `crypto_authenc_*()` implementation to `core/crypto/crypto.c`, implemented by using the new functions above.

I suppose the best would be to solve the `crypto_authenc_*()` problem in this pull request to avoid messing with crypto interface little by little (or more truthfully much by much).

Another thing is the `core/include/tee/tee_cryp_provider.h` file, it should probably be renamed to `core/include/crypto/api.h` or something similar.

Should I move forward as proposed above? Comments/thoughts?

### CODE REVIEW

**Comment by jforissier** [comment]
TL;DR: agreed with everything,

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Indeed, replacing `crypto_ops` is not sufficient in itself to address the mixed GCM/CCM implementation issue, but it is a step in the right direction anyway (unless someone chimes in and says we need to keep function pointers, but why would we need that?). We do need a lower-level API than the current `crypto_ops`, and I am perfectly fine with your suggestion (1) and (2). I think it belongs in this PR because then you can rebase 1908 on this.

I also agree with renaming `core/include/tee/tee_cryp_provider.h`. I would suggest `core/include/crypto/crypto.h` rather than `core/include/crypto/api.h`, because it's more meaningful when the full path is not mentioned ("crypto.h" rather than "api.h"). Not that it matters much however.

**Comment by jenswi-linaro** [comment]
Sounds good. :-)

`core/include/crypto/crypto.h`, it is then. I'll squash and apply the tag first to make this first step clear.

**Comment by jenswi-linaro** [comment]
Tag applied

**Comment by jbech-linaro** [comment]
So, besides the code we have some more places that needs to be fixed due to this
* https://github.com/OP-TEE/optee_os/blob/master/documentation/porting_guidelines.md#6-hardware-crypto-ip
* https://github.com/OP-TEE/optee_os/blob/master/documentation/crypto.md
* https://github.com/OP-TEE/optee_website/tree/master/faq#can-i-use-my-own-hardware-ip-for-crypto-acceleration

File: core/crypto/crypto.c [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+#include <compiler.h>
+#include <kernel/panic.h>
+#include <tee/tee_cryp_provider.h>
+
+#if !defined(_CFG_CRYPTO_WITH_HASH)
+TEE_Result crypto_hash_get_ctx_size(uint32_t algo __unused,
+				    size_t *size __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
```
Comment (etienne-lms) [inline core/crypto/crypto.c]: `TEE_ERROR_NOT_IMPLEMENTED` instead, for consistency ?
Only hash function use here `TEE_ERROR_NOT_SUPPORTED`. Others uses `TEE_ERROR_NOT_IMPLEMENTED`.

File: core/include/tee/tee_cryp_provider.h [hunk 2]
```diff
@@ -198,102 +185,82 @@ struct ecc_keypair {
...
-	struct cipher_ops cipher;
-	struct mac_ops mac;
-	struct authenc_ops authenc;
-	struct acipher_ops acipher;
-	struct bignum_ops bignum;
-};
+/*
+ * Key allocation functions
+ * Allocate the bignum's inside a key structure.
+ * TEE core will later use bignum.free().
```
Comment (etienne-lms) [inline core/include/tee/tee_cryp_provider.h]: minor: s/bignum.free/crypto_bignum_free/

**Comment by jforissier** [comment]
@jbech-linaro correct for the first two, but the last one (the faq) is OK I think.

File: core/tee/tee_rpmb_fs.c [hunk 3]
```diff
@@ -1178,9 +1167,8 @@ static TEE_Result tee_rpmb_init(uint16_t dev_id)
 		memcpy(rpmb_ctx->cid, dev_info.cid, RPMB_EMMC_CID_SIZE);
 
 		if ((rpmb_ctx->hash_ctx_size == 0) &&
-		    (crypto_ops.mac.get_ctx_size(
-			    TEE_ALG_HMAC_SHA256,
-			    &rpmb_ctx->hash_ctx_size))) {
+		    (crypto_mac_get_ctx_size(TEE_ALG_HMAC_SHA256,
+					     &rpmb_ctx->hash_ctx_size))) {
```
Comment (etienne-lms) [inline core/tee/tee_rpmb_fs.c]: minor: can remove the parentheses around `crypto_mac_get_ctx_size()`.

**Comment by jbech-linaro** [comment]
> @jbech-linaro correct for the first two, but the last one (the faq) is OK I think.

You're probably right, but is it really an abstraction layer anymore? Maybe?

**Comment by jenswi-linaro** [comment]
Yes, it's still an abstraction layer.

File: core/crypto/crypto.c [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <compiler.h>
+#include <kernel/panic.h>
+#include <tee/tee_cryp_provider.h>
+
+#if !defined(_CFG_CRYPTO_WITH_HASH)
```
Comment (jforissier) [inline core/crypto/crypto.c]: How about making the symbols `__weak` instead? It looks like we could get rid of all the #ifdef-ery in this file.

File: core/crypto/crypto.c (hunk 1 above)
Comment (jenswi-linaro) [reply core/crypto/crypto.c]: Thanks, I'll replace all those.

File: core/crypto/crypto.c (hunk 4 above)
Comment (jenswi-linaro) [reply core/crypto/crypto.c]: I was thinking the same first, but the problem is that some functions are supposed to be provided in groups and that's not possible with weak functions.

**Comment by jenswi-linaro** [comment]
Addressed comments, moved `crypto_authenc_*()` as agreed.

**Comment by jforissier** [comment]
Section "How to add a new crypto implementation" in `documentation/crypto.md` needs updating, too.
Other than that LGTM, thanks.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: documentation/crypto.md [hunk 5]
```diff
@@ -51,68 +51,68 @@ All cryptography-related system calls are declared in
...
+        return TEE_ERROR_NOT_SUPPORTED;
+}
+...
+#endif /*_CFG_CRYPTO_WITH_HASH*/
+
+/*
+ * Implementation of crypto_authenc_*() due to divided implementation,
+ * AES-GCM is normally using an OP-TEE internal implementation while
+ * AES-CCM is using the LibTomCrypt implementation.
+ */
```
Comment (etienne-lms) [inline documentation/crypto.md]: remove this comment

File: documentation/crypto.md [hunk 6]
```diff
@@ -51,68 +51,68 @@ All cryptography-related system calls are declared in
...
  */
 
-struct crypto_ops crypto_ops = {
-	.name = "LibTomCrypt provider",
-	.init = tee_ltc_init,
+#if !defined(_CFG_CRYPTO_WITH_HASH)
+TEE_Result crypto_hash_get_ctx_size(uint32_t algo __unused,
+                                    size_t *size __unused)
+{
+        return TEE_ERROR_NOT_SUPPORTED;
```
Comment (etienne-lms) [inline documentation/crypto.md]: s/TEE_ERROR_NOT_SUPPORTED/TEE_ERROR_NOT_IMPLEMENTED/

**Comment by jenswi-linaro** [comment]
Update

File: documentation/crypto.md [hunk 7]
```diff
@@ -163,22 +157,18 @@ return **TEE_ERROR_NOT_IMPLEMENTED**).
...
+  code that will be used regardless of which crypto provider is in use.
+  How we are dealing with AES-GCM in [core/crypto](../core/crypto) could
+  serve as an example.
 - Avoid modifying [tee_svc_cryp.c](../core/tee/tee_svc_cryp.c). It should not be
   needed.
-- Your own **struct crypto_ops crypto_ops = ...** should be defined in a file at
-  the top level of your new directory.
-- Although not all pointers in **crypto_ops** need to be defined, all are
-  required for compliance to the GlobalPlatform specification.
+- Although not crypto families need to be defines, all are required for
```
Comment (etienne-lms) [inline documentation/crypto.md]: "... not **all** crypto families ..." ?

File: core/crypto/crypto.c [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+		if (*dst_len < src_len)
+			return TEE_ERROR_SHORT_BUFFER;
+		*dst_len = src_len;
+		return crypto_aes_ccm_update_payload(ctx, mode, src_data,
+						     src_len, dst_data);
+#endif
+#if defined(CFG_CRYPTO_GCM)
+	case TEE_ALG_AES_GCM:
+		if (*dst_len < src_len)
+			return TEE_ERROR_SHORT_BUFFER;
```
Comment (etienne-lms) [inline core/crypto/crypto.c]: should set *dst_len = src_len; before returning `TEE_ERROR_SHORT_BUFFER`.
here and below, and in few below functions.

Comment (jenswi-linaro) [reply core/crypto/crypto.c]: Thanks

**Comment by jenswi-linaro** [comment]
Update

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960, GP)`

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 9]
```diff
@@ -2546,293 +2532,288 @@ struct tee_ccm_state {
...
+}
+
+TEE_Result crypto_aes_ccm_dec_final(void *ctx, const uint8_t *src_data,
+				    size_t len, uint8_t *dst_data,
+				    const uint8_t *tag, size_t tag_len)
+{
+	TEE_Result res = TEE_ERROR_BAD_STATE;
+	struct tee_ccm_state *ccm = ctx;
+	int ltc_res;
+	uint8_t dst_tag[TEE_xCM_TAG_MAX_LENGTH];
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: minor: s/`TEE_xCM_TAG_MAX_LENGTH`/`TEE_CCM_TAG_MAX_LENGTH`/ here and line 2665.

(definition of `TEE_xCM_TAG_MAX_LENGTH` could be removed.

File: core/lib/libtomcrypt/src/tee_ltc_provider.c [hunk 10]
```diff
@@ -2546,293 +2532,288 @@ struct tee_ccm_state {
...
 	TEE_Result res = TEE_ERROR_BAD_STATE;
-#if defined(CFG_CRYPTO_CCM)
-	struct tee_ccm_state *ccm;
-#endif
-#if defined(CFG_CRYPTO_GCM)
-	struct tee_gcm_state *gcm;
-#endif
+	struct tee_gcm_state *gcm = ctx;
 	int ltc_res;
 	uint8_t dst_tag[TEE_xCM_TAG_MAX_LENGTH];
```
Comment (etienne-lms) [inline core/lib/libtomcrypt/src/tee_ltc_provider.c]: minor: s/`TEE_xCM_TAG_MAX_LENGTH`/`TEE_GCM_TAG_MAX_LENGTH`/ here and line 2823.

**Comment by jenswi-linaro** [comment]
Addressed comments
@jforissier and @etienne-lms are you still OK with this? If so I'll squash and rebase.

**Comment by etienne-lms** [comment]
fine for me

**Comment by jenswi-linaro** [comment]
Squashed, rebased and tags applied.
